### PR TITLE
Sanitize dialog text-body.

### PR DESCRIPTION
**Pull request by Carreau**
Mitigate 7486, could probably be improved, but
treat text as pure text, but assume that when received a dom
element it actually come from a secure source.

Sanitizing all we get, including Dom Element break json-metadata editor
and a few others.

### CODE REVIEW

File: IPython/html/static/base/js/dialog.js [hunk 1]
```diff
@@ -20,6 +21,9 @@ define([
         var dialog_content = $("<div/>")
             .addClass("modal-content")
             .appendTo(dialog);
+        if(typeof(options.body) === 'string'){
+            options.body = $("<p/>").text(security.sanitize_html(options.body))
```
Comment (minrk) [inline IPython/html/static/base/js/dialog.js]: There's no reason to use `.text` *and* `sanitize_html`, since `.text` escapes everything.

Comment (minrk) [inline IPython/html/static/base/js/dialog.js]: But I agree that using `.text` is a good idea, here.

Comment (Carreau) [inline IPython/html/static/base/js/dialog.js]: Don't see what you are talking about. ( :eyes: wistle while looking away )

File: IPython/html/static/base/js/dialog.js [hunk 2]
```diff
@@ -32,7 +35,7 @@ define([
                 ).append(
                     $("<h4/>")
                         .addClass('modal-title')
-                        .text(options.title || "")
+                        .text(options.title || "No Title")
```
Comment (minrk) [inline IPython/html/static/base/js/dialog.js]: I think actually having no title is better than giving the dialog a title called "No Title"

File: IPython/html/static/base/js/dialog.js [hunk 3]
```diff
@@ -20,6 +20,9 @@ define([
         var dialog_content = $("<div/>")
             .addClass("modal-content")
             .appendTo(dialog);
+        if(typeof(options.body) === 'string'){
+            options.body = $("<p/>").text(options.body)
```
Comment (jdfreder) [inline IPython/html/static/base/js/dialog.js]: Doesn't this break the rename function of the savewidget ?

**Comment by jdfreder** [comment]
I'm not sure why we need to sanitize the dialog method.  i think this will break the logic we use to rename notebooks, where we pass an input control into the body to make the dialog accept text.  I'm planning on using this same technique elsewhere in IPython, and I think it's an appropriate use of the dialog helper function (I'd rather not have to re-implement this everywhere I need it just because the base only accepts text).  Instead, maybe it's better sanitize the content before it is passed into this method.

File: IPython/html/static/base/js/dialog.js (hunk 3 above)
Comment (jdfreder) [inline IPython/html/static/base/js/dialog.js]: Oh I see, JQuery selectors pass over this.

Comment (minrk) [inline IPython/html/static/base/js/dialog.js]: Yup, it only sanitizes string dialogs. HTML dialogs aren't touched, and won't be.

**Comment by minrk** [comment]
> maybe it's better sanitize the content before it is passed into this method.

yes, definitely. It is absolutely up to the builder of the dialog HTML to make sure it's safe.

**Comment by Carreau** [comment]
if you pass it a dom object it will behave as expected. so you can alcually say `modal({body:$('p').append(text)})` it will work. The problem is that in some places, I had the dialog sow error/response message from server, which is a potential security flaw to inject js. It is better to sanitize by default everywhere and be explicit when you want safety off than the other way around.

**Comment by jdfreder** [comment]
> The problem is that in some places, I had the dialog sow error/response message from server, which is a potential security flaw to inject js. It is better to sanitize by default everywhere and be explicit when you want safety off than the other way around.

I disagree with this.  This dialog method should be treated like a `.html` method.  The error messages should be sanitized before being passed to the dialog method.

Alternatively, you could create a convenience flag on this method, i.e. `sanitize: true` which triggered sanitation to happen.  I think that would be better than having sanitation triggered implicitly by argument type.

File: IPython/html/static/base/js/dialog.js (hunk 2 above)
Comment (jdfreder) [inline IPython/html/static/base/js/dialog.js]: Agreed

**Comment by Carreau** [comment]
that would be fine with me also, but sanitation should then be true by default, that was my first attempt at this PR. Though, the sanitizing library does work only on string, so this path would be more invasive.

**Comment by minrk** [comment]
> error/response message from server, which is a potential security flaw to inject js

Since the vulnerability would have to be in the same code that serves the sanitization js, I'm not sure I see the issue here.

A disadvantage of sanitizing string bodies means that you have to be sure that the caller  *doesn't* sanitize the string body, otherwise you will get double-escapes. I still think this probably makes sense to do, and we can make a pass on the HTML inputs, to see if there are issues. For instance, the error messages from the server all pass through a single ajax error utility, which should probably call `escape_html(message)`, so that we know that `error.message` is safe text everywhere we use it.

**Comment by Carreau** [comment]
I'm more concern by the fact that we expose this as an API that is used by extensions. 

I know extensions are not our problem and that security is their problem, but providing an unsecure default has historically always be a bad choice.

**Comment by minrk** [comment]
I think it's easy to be surprised by the plain-text being able to execute HTML, which is why I still think this PR is a good idea. If they are passing HTML, I don't think it's an issue *at all*, because they could be putting that same HTML on the page in any way they want, and it would always be applied as-is. That's why I think this PR, plus a pass on our own HTML dialogs, closes #7486.

**Comment by jdfreder** [comment]
@Carreau what do you think about

```js
if (options.sanitize===undefined || options.sanitize) {
    options.body = $("<p/>").text(options.body)
}
```

?

**Comment by Carreau** [comment]
`options.sanitize===undefined || options.sanitize` -> `options.sanitize!==false` to avoid things like `sanitize='sdkjgb'` to pass through would suite me.

**Comment by Carreau** [comment]
Updated to @jdfreder comment. 
`p.text()` will be used only if string is passed and sanitize is not set to `false`.

**Comment by jdfreder** [comment]
Thanks! :)

**Comment by takluyver** [comment]
From the dev meeting: we're happy with what this does at present (escape strings passed to dialog, don't sanitise pre-constructed DOM elements, because whatever constructs them is responsible for doing so safely).

**Comment by takluyver** [comment]
Travis passed. Who wants to do final review and/or press the button?

**Pull request merged**